File: code_samples/go/design_patterns_decorator.go
package main

import (
    "fmt"
)

// Design Patterns: Decorator
// AI/ML Training Sample

type Decorator struct {
    Data string
}

func NewDecorator() *Decorator {
    return &Decorator{
        Data: "",
    }
}

func (s *Decorator) Process(input string) {
    s.Data = input
}

func (s *Decorator) Validate() bool {
    return len(s.Data) > 0
}

func (s *Decorator) GetData() string {
    return s.Data
}

func main() {
    instance := NewDecorator()
    instance.Process("example")
    fmt.Printf("Data: %s\n", instance.GetData())
    fmt.Printf("Valid: %v\n", instance.Validate())
}

File: code_samples/scala/design_patterns_decorator.scala
/**
 * Design Patterns: Decorator
 * AI/ML Training Sample
 */

class Decorator {
  private var data: String = ""
  
  def process(input: String): Unit = {
    data = input
  }
  
  def getData: String = data
  
  def validate: Boolean = data.nonEmpty
}

object DecoratorExample {
  def main(args: Array[String]): Unit = {
    val instance = new Decorator()
    instance.process("example")
    println(s"Data: ${instance.getData}")
    println(s"Valid: ${instance.validate}")
  }
}
